mrmustard/lab_dev/states/base.py

```diff
@@ -843,11 +843,39 @@ def random(cls, modes, m=None, max_r=1.0):
         max_idx = max(modes)
 
         ancilla = list(range(max_idx + 1, max_idx + m + 1))
-        full_wires = modes + ancilla
+        full_wires = list(modes) + ancilla
 
         psi = Ket.random(full_wires, max_r)
         return psi[modes]
 
+    @property
+    def is_positive(self) -> bool:
+        r"""
+        Whether this DM is a positive operator.
+        """
+        batch_dim = self.representation.A.shape[0]
+        if batch_dim > 1:
+            raise ValueError(
+                "Physicality conditions are not implemented for batch dimension larger than 1."
+            )
+        A = self.representation.A[0]
+        m = A.shape[-1] // 2
+        gamma_A = A[:m, m:]
+
+        if (
+            math.real(math.norm(gamma_A - math.conj(gamma_A.T))) > 1e-8
+        ):  # checks if gamma_A is Hermitian
+            return False
+
+        return all(math.real(mu) >= 0 for mu in math.eigvals(gamma_A))
```

Comment:
```suggestion
	return all(math.real(math.eigvals(gamma_A)) >= 0)
```
Although not sure does `all` work with tensorflow? I'd assume yes but just making sure

Reply:
Nice.. It passed the tensorflow tests, so I guess that's working.